## Experimentation with Different Models


## Introduction to YOLOv8 for Object Detection in Inclement Weather

Object detection in challenging conditions, such as inclement weather, requires a robust and efficient model capable of recognizing objects with high precision despite visual impairments caused by rain, fog, snow, or low light. The YOLO (You Only Look Once) series has been at the forefront of object detection tasks due to its speed and accuracy, and the latest iteration, YOLOv8, brings further improvements in both performance and efficiency.

YOLOv8 stands out for its ability to process images in real-time with impressive accuracy, making it an ideal candidate for applications requiring quick and reliable object detection in adverse weather conditions. Its architecture is designed to handle the complexities and variability of real-world scenarios, which is critical for ensuring safety and operational efficiency in environments affected by inclement weather.

We will experiment with the three available variants of the YOLOv8 model: nano, small, and medium. These models vary in size and computational requirements, allowing us to assess which variant offers the best balance between speed and accuracy for our prototype. The **nano** model, being the lightest, promises the fastest inference times but at potentially lower accuracy. The **small** model offers a middle ground, aiming for a balance between performance and speed. Lastly, the **medium** model provides the highest accuracy at the cost of increased computational load. Our goal is to identify the most suitable model for object detection in inclement weather, ensuring that our prototype can operate effectively under challenging conditions.


## 1. Data Label Fromat

### 1.1. Original Annotation Format to KITTI Format

The filtermoosekitti.py script, sourced from the asvath/cadcd repository, plays a crucial role in ensuring the integrity and standardization of our dataset annotations. Developed by the original author, this script filters annotations based on a minimum of 20 lidar points, effectively eliminating potential noise or outliers from the ground truth data. This methodical approach not only aligns annotations with the KITTI format but also incorporates essential details such as truncation, observation angle, and pixel coordinates. The script's commitment to maintaining a standardized layout enhances the overall robustness of the dataset, making it well-suited for experimentation with various object detection models. For further reference, you can find the script in the asvath/cadcd repository.

Then, we rename the images inside each Image_## folder in the 'processedmoose' so that the images' names matches the annotation files' numenclature.


In [ ]:
import os

# Define the parent directory where the 'image_00', 'image_01', ..., 'image_07' folders are located
parent_directory = 'C:\\Users\\rafae\\Documents\\Capstone\\autonomoose\\processedmoose'  

# Loop through each subfolder (image_00, image_01, ..., image_07)
for i in range(8):
    subfolder_name = 'image_{:02d}'.format(i)
    subfolder_path = os.path.join(parent_directory, subfolder_name, 'data') 

    # Loop through each image in the subfolder
    for j in range(98):
        old_image_name = '000000{:04d}.png'.format(j)
        new_image_name = '000000{:04d}.png'.format(i * 100 + j)

        old_image_path = os.path.join(subfolder_path, old_image_name)
        new_image_path = os.path.join(subfolder_path, new_image_name)

        # Rename the image
        os.rename(old_image_path, new_image_path)

Great, we have remaned all images in the folers image_##: {old_image_name} -> {new_image_name}
    
Image_00 > 0000000002.png -> 0000000002.png, ..., 00000000099.png -> 00000000099.png
    
Image_01 > 0000000002.png -> 0000000102.png, ..., 00000000099.png -> 00000000199.png
    
...
    
Image_07 > 0000000002.png -> 0000000702.png, ..., 00000000099.png -> 00000000799.png

### 1.2. KITTI to VOC Dataset Converter

Now, we can transform the the data from KITTI format to VOC format using the Visual Object Dataset Converter from https://github.com/umautobots/vod-converter/blob/master/README.md 

Run the following command in the terminal to convert datasets from the KITTI format to VOC:

```bash
python vod-converter/vod_converter/main.py --from kitti --from-path cadc_data --to voc --to-path mydata-voc


### 1.3. Dataset Preprocessing with Roboflow

Now that all images share the same names as their corresponding annotation files, we are ready to upload this dataset to Roboflow. Roboflow plays a crucial role in facilitating the data formatting step. With the images and annotations in VOC format, Roboflow can read this format and convert it into other widely used standards, such as YOLO. This eliminates the need for manual coding during format changes. Also, Roboflow also has augmentation settings that we can use when processing the dataset.

We successfully uploaded the 198 images along with their corresponding annotation files to Roboflow. During this process, Roboflow automatically filtered out any annotations with invalid data, such as negative height or width, empty annotations, and more. As a result, we now have a dataset containing 198 annotated images, 100 with no snow, 98 with snow-covered roads, featuring **three classes**: 

**1. Car**

**2. Truck**

**3. Person** (*Pedestrian* was renamed to *Person* since we're using the pretrained model on COCO dataset which has the class *Person* not *Pedestrian*)

Subsequently, Roboflow partitioned the dataset into training, validation, and test sets.

## 2. YOLOv8 

### 2.1. Experiment Design and Model Selection for Cloudy and Snowy Conditions

In our endeavor to develop a robust object detection prototype, we are conducting an experiment using a dataset that initially comprises 100 images captured under cloudy weather conditions. Recognizing the importance of versatility in real-world applications, we plan to augment this dataset with an additional 98 images featuring snow conditions. This expansion will enable us to assess the prototype's performance across a broader spectrum of inclement weather scenarios. Our experiment is structured to run for 50 epochs, providing a thorough training period for the models under consideration.

### Dataset Composition Analysis
The initial focus on cloudy weather aims to address the specific challenges posed by overcast skies and diminished lighting. The subsequent inclusion of snow images is strategic, introducing complexities such as occlusion and altered landscape contrast, which are critical for testing the model's adaptability and accuracy in diverse environments. 

Aditionally, the dataset encompasses a total of 720 labels, distributed across two categories:

- **Cars**: 599 labels, the majority of our dataset, focusing on vehicular detection.
- **Trucks**: 121 labels, introducing larger vehicle detection challenges.

The modest size of our dataset, while potentially limiting, is intentionally chosen to ensure focused, rapid initial experimentation. We acknowledge the risks associated with a small dataset, such as overfitting, and plan to implement data augmentation and appropriate regularization techniques to mitigate these concerns. This dataset serves as a pivotal foundation for establishing baseline performance metrics and understanding the nuanced effects of weather on detection capabilities.

The dataset used in this notebook can be downloaded from Roboflow using the badge bellow.


<a href="https://universe.roboflow.com/machine-learning-engineer-bootcamp/object_detect-bh8cz">
    <img src="https://app.roboflow.com/images/download-dataset-badge.svg"></img>
</a>

### Benchmarking Strategy with YOLOv8 Variants
Our benchmarking strategy encompasses evaluating three variants of the YOLOv8 model: nano, small, and medium. This selection is guided by the need to balance computational efficiency with detection accuracy:

- **YOLOv8 nano**: Tested for its speed and efficiency, particularly valuable for real-time applications, albeit with an anticipated compromise in accuracy under complex weather conditions.
- **YOLOv8 small**: Expected to strike an optimal balance, potentially offering superior accuracy to the nano model with manageable computational demands, making it a strong candidate for the prototype.
- **YOLOv8 medium**: While the most computationally intensive, it will be benchmarked for its maximum achievable accuracy, setting an aspirational standard for detection performance in adverse weather.

### Future Directions and Model Choice
The inclusion of snow images and the evaluation on the chosen YOLOv8 model, together with fine-tuning, will inform our decision on the most suitable model for our prototype. The aim is to select a model that demonstrates high accuracy in diverse weather conditions while maintaining computational efficiency. 

This initial phase of experimentation and model selection lays the groundwork for subsequent prototype iterations. With the chosen model, we intend to further refine our approach by expanding the dataset, incorporating more varied weather conditions, and optimizing model parameters to enhance detection capabilities in real-world scenarios. The insights gained will drive the continuous improvement of our object detection solution, ensuring reliability and effectiveness in changing environmental conditions.

### 2.2. Experimenting with Different Models

In [8]:
from ultralytics import YOLO
import ultralytics
ultralytics.checks()

Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Setup complete ✅ (4 CPUs, 15.8 GB RAM, 233.7/953.3 GB disk)


### YOLOv8n (Nano)

In [9]:
model = YOLO('yolov8n.pt')  # load a pretrained model (recommended for training)

In [10]:
# Train the model
results = model.train(data='data.yaml', epochs=50, imgsz=640, plots=True)

New https://pypi.org/project/ultralytics/8.1.11 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=data.yaml, epochs=50, patience=50, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train7, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, show=False, save_frames=False, save_txt=False, save_conf=False, save_c

train: Scanning C:\Users\rafae\Documents\Capstone\object_detect-1\train\labels.cache... 63 images, 11 backgrounds, 0 corrupt: 100%|██████████| 63/63 [00:00<?, ?it/s]
val: Scanning C:\Users\rafae\Documents\Capstone\object_detect-1\valid\labels.cache... 19 images, 1 backgrounds, 0 corrupt: 100%|██████████| 19/19 [00:00<?, ?it/s]

Plotting labels to runs\detect\train7\labels.jpg... 


optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs\detect\train7
Starting training for 50 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/50         0G      2.277      4.127      1.452        121        640: 100%|██████████| 4/4 [01:05<00:00, 16.46s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.05s/it]

                   all         19        151    0.00532      0.203     0.0293     0.0151



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/50         0G      2.248      3.728      1.388        231        640: 100%|██████████| 4/4 [01:15<00:00, 18.93s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.19s/it]

                   all         19        151    0.00966      0.383      0.106     0.0539



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/50         0G       2.14      2.953      1.271        155        640: 100%|██████████| 4/4 [01:06<00:00, 16.69s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.70s/it]

                   all         19        151      0.011      0.364       0.14     0.0562



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/50         0G      1.867      2.434      1.198        118        640: 100%|██████████| 4/4 [01:15<00:00, 18.88s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/1 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 1.450s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.27s/it]

                   all         19        151     0.0109        0.4      0.248     0.0975



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/50         0G      1.929      2.189      1.249        189        640: 100%|██████████| 4/4 [01:08<00:00, 17.12s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.36s/it]

                   all         19        151     0.0118      0.474      0.267     0.0878



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/50         0G      1.946      2.106      1.239        213        640: 100%|██████████| 4/4 [01:06<00:00, 16.56s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.16s/it]

                   all         19        151     0.0101      0.438       0.24      0.103



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/50         0G      1.819      1.971      1.234        133        640: 100%|██████████| 4/4 [01:04<00:00, 16.05s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.58s/it]

                   all         19        151     0.0102      0.434      0.245      0.134



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/50         0G      1.811      1.848      1.222        194        640: 100%|██████████| 4/4 [01:03<00:00, 15.99s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.57s/it]

                   all         19        151    0.00928      0.408      0.213      0.108



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/50         0G      1.771      1.877      1.191        129        640: 100%|██████████| 4/4 [01:09<00:00, 17.35s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.81s/it]

                   all         19        151          1     0.0392      0.186     0.0902



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/50         0G      1.821      1.908      1.227         91        640: 100%|██████████| 4/4 [01:09<00:00, 17.28s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.62s/it]

                   all         19        151     0.0133      0.532      0.165     0.0804



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/50         0G      1.949      1.769      1.258        171        640: 100%|██████████| 4/4 [01:04<00:00, 16.14s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.22s/it]

                   all         19        151     0.0108      0.431      0.109     0.0532



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/50         0G      1.816       1.81      1.241        176        640: 100%|██████████| 4/4 [01:05<00:00, 16.38s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.81s/it]

                   all         19        151     0.0128       0.52     0.0622     0.0347



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/50         0G      1.809       1.81      1.219        193        640: 100%|██████████| 4/4 [01:02<00:00, 15.70s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.44s/it]

                   all         19        151     0.0134      0.517     0.0889     0.0523



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/50         0G      1.732      1.771       1.22        153        640: 100%|██████████| 4/4 [01:04<00:00, 16.10s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.03s/it]

                   all         19        151     0.0149      0.549     0.0971     0.0574



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/50         0G      1.703      1.718      1.223        154        640: 100%|██████████| 4/4 [01:02<00:00, 15.54s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.30s/it]

                   all         19        151     0.0449      0.335      0.135     0.0754



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/50         0G      1.691       1.72      1.213        127        640: 100%|██████████| 4/4 [01:07<00:00, 17.00s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.64s/it]

                   all         19        151      0.407      0.102      0.141     0.0815



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/50         0G       1.77      1.707      1.183        185        640: 100%|██████████| 4/4 [01:08<00:00, 17.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.12s/it]

                   all         19        151     0.0176        0.6      0.175     0.0967



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/50         0G      1.761      1.748      1.176        143        640: 100%|██████████| 4/4 [01:10<00:00, 17.59s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.75s/it]

                   all         19        151      0.406      0.311      0.277      0.152



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/50         0G      1.693      1.664      1.164        166        640: 100%|██████████| 4/4 [01:12<00:00, 18.15s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.13s/it]

                   all         19        151      0.712      0.156      0.333      0.179



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/50         0G      1.703      1.721      1.201        122        640: 100%|██████████| 4/4 [01:05<00:00, 16.49s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.17s/it]

                   all         19        151      0.909      0.108      0.321      0.171



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/50         0G      1.646      1.627      1.182        153        640: 100%|██████████| 4/4 [01:08<00:00, 17.22s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.11s/it]

                   all         19        151          1     0.0433      0.203      0.111



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/50         0G      1.609      1.712      1.155        164        640: 100%|██████████| 4/4 [01:14<00:00, 18.66s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:12<00:00, 12.03s/it]

                   all         19        151      0.487     0.0533      0.301      0.162



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/50         0G       1.62      1.618      1.183        110        640: 100%|██████████| 4/4 [01:15<00:00, 18.88s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/1 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 1.450s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:12<00:00, 12.57s/it]

                   all         19        151      0.717      0.194      0.375      0.202



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/50         0G      1.551      1.526      1.172        134        640: 100%|██████████| 4/4 [01:24<00:00, 21.05s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:11<00:00, 11.63s/it]

                   all         19        151      0.486      0.243      0.373      0.209



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/50         0G      1.594      1.583       1.14        182        640: 100%|██████████| 4/4 [01:25<00:00, 21.47s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:13<00:00, 13.34s/it]

                   all         19        151      0.531      0.289      0.394      0.219



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/50         0G      1.618      1.552      1.155        138        640: 100%|██████████| 4/4 [01:23<00:00, 20.75s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.99s/it]

                   all         19        151      0.476      0.278      0.391      0.206



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/50         0G      1.627      1.498      1.154        169        640: 100%|██████████| 4/4 [01:34<00:00, 23.63s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:12<00:00, 12.46s/it]

                   all         19        151      0.557      0.315      0.368      0.209



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/50         0G      1.632      1.536       1.16        131        640: 100%|██████████| 4/4 [01:20<00:00, 20.15s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:12<00:00, 12.13s/it]

                   all         19        151       0.54      0.333       0.39      0.229



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/50         0G      1.595      1.558      1.144        156        640: 100%|██████████| 4/4 [01:04<00:00, 16.09s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.20s/it]

                   all         19        151      0.526      0.344      0.428      0.257



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/50         0G      1.477      1.426      1.129        168        640: 100%|██████████| 4/4 [01:06<00:00, 16.58s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.44s/it]

                   all         19        151      0.526      0.434      0.471      0.262



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      31/50         0G      1.535      1.421      1.108        181        640: 100%|██████████| 4/4 [01:05<00:00, 16.32s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.15s/it]

                   all         19        151      0.521      0.486      0.516      0.287



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      32/50         0G      1.478      1.404      1.128        147        640: 100%|██████████| 4/4 [01:03<00:00, 15.78s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:11<00:00, 11.46s/it]

                   all         19        151       0.49      0.429      0.475      0.278



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      33/50         0G      1.459       1.38      1.113        163        640: 100%|██████████| 4/4 [01:06<00:00, 16.62s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.56s/it]

                   all         19        151      0.453      0.465       0.46      0.273



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      34/50         0G      1.484      1.404      1.091        103        640: 100%|██████████| 4/4 [01:00<00:00, 15.14s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.33s/it]

                   all         19        151      0.477      0.513       0.46       0.27



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      35/50         0G      1.437      1.434      1.146        187        640: 100%|██████████| 4/4 [01:27<00:00, 21.86s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:11<00:00, 11.61s/it]

                   all         19        151      0.511      0.509      0.481      0.282



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      36/50         0G       1.43      1.368      1.087        112        640: 100%|██████████| 4/4 [01:28<00:00, 22.18s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:11<00:00, 11.87s/it]

                   all         19        151      0.575      0.486      0.512      0.311



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      37/50         0G      1.387      1.342      1.112        115        640: 100%|██████████| 4/4 [01:23<00:00, 20.89s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.27s/it]

                   all         19        151        0.6       0.47      0.522      0.319



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      38/50         0G      1.419      1.335      1.098        129        640: 100%|██████████| 4/4 [01:16<00:00, 19.02s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:11<00:00, 11.12s/it]

                   all         19        151      0.571      0.504      0.545      0.321



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      39/50         0G      1.399      1.288      1.118        131        640: 100%|██████████| 4/4 [01:21<00:00, 20.32s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.78s/it]

                   all         19        151      0.562      0.538      0.553      0.334



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      40/50         0G      1.388      1.317      1.066        128        640: 100%|██████████| 4/4 [01:22<00:00, 20.52s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.40s/it]

                   all         19        151      0.572      0.521      0.557      0.334


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      41/50         0G      1.351      1.481      1.069        120        640: 100%|██████████| 4/4 [01:13<00:00, 18.33s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.46s/it]

                   all         19        151      0.581      0.506       0.56      0.333



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      42/50         0G      1.323      1.366      1.056         97        640: 100%|██████████| 4/4 [01:00<00:00, 15.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.25s/it]

                   all         19        151      0.569      0.491      0.555      0.327



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      43/50         0G      1.307      1.268      1.071         82        640: 100%|██████████| 4/4 [01:00<00:00, 15.10s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.50s/it]

                   all         19        151      0.491      0.535       0.56      0.334



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      44/50         0G      1.323       1.33      1.056         82        640: 100%|██████████| 4/4 [01:03<00:00, 15.84s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.26s/it]

                   all         19        151      0.497       0.53      0.559      0.341



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      45/50         0G        1.3      1.376      1.061        109        640: 100%|██████████| 4/4 [01:02<00:00, 15.54s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.38s/it]

                   all         19        151      0.535      0.581      0.569      0.351



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      46/50         0G      1.273      1.331      1.057        101        640: 100%|██████████| 4/4 [01:00<00:00, 15.18s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.25s/it]

                   all         19        151      0.554       0.63      0.589      0.361



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      47/50         0G      1.275      1.302      1.074        104        640: 100%|██████████| 4/4 [01:05<00:00, 16.30s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:09<00:00,  9.16s/it]

                   all         19        151      0.573      0.621      0.594      0.369



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      48/50         0G      1.247      1.298      1.045        130        640: 100%|██████████| 4/4 [00:58<00:00, 14.61s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.77s/it]

                   all         19        151      0.568       0.61      0.603      0.367



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      49/50         0G      1.223      1.295      1.031         99        640: 100%|██████████| 4/4 [01:03<00:00, 15.93s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:07<00:00,  7.80s/it]

                   all         19        151      0.552      0.615      0.615      0.371



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      50/50         0G       1.22      1.234      1.054         97        640: 100%|██████████| 4/4 [01:01<00:00, 15.39s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.89s/it]

                   all         19        151      0.561      0.591      0.621      0.379



50 epochs completed in 1.145 hours.
Optimizer stripped from runs\detect\train7\weights\last.pt, 6.3MB
Optimizer stripped from runs\detect\train7\weights\best.pt, 6.3MB

Validating runs\detect\train7\weights\best.pt...
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 168 layers, 3006038 parameters, 0 gradients, 8.1 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:08<00:00,  8.17s/it]


                   all         19        151      0.556      0.581      0.621      0.379
                 truck         19         26       0.46      0.394      0.471      0.275
                   car         19        125      0.652      0.768      0.771      0.482
Speed: 5.8ms preprocess, 357.6ms inference, 0.0ms loss, 50.3ms postprocess per image
Results saved to runs\detect\train7


In [12]:
# Train the model
#results = model.train(data='data.yaml', epochs=50, imgsz=640, plots=True)

metrics = model.val()
metrics.box.map    # map50-95
metrics.box.map50  # map50
metrics.box.map75  # map75
metrics.box.maps   # a list contains map50-95 of each category

Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 168 layers, 3006038 parameters, 0 gradients, 8.1 GFLOPs


val: Scanning C:\Users\rafae\Documents\Capstone\object_detect-1\valid\labels.cache... 19 images, 1 backgrounds, 0 corrupt: 100%|██████████| 19/19 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:07<00:00,  3.62s/it]


                   all         19        151      0.556      0.581      0.621      0.379
                 truck         19         26       0.46      0.394      0.471      0.275
                   car         19        125      0.652      0.768      0.771      0.482
Speed: 10.0ms preprocess, 316.2ms inference, 0.0ms loss, 34.9ms postprocess per image
Results saved to runs\detect\train72


array([      0.275,     0.48229])

![Description or alt text for the image](runs\detect\train7\results.png "Optional title")

### YOLOv8s (Small)

In [4]:
model1 = YOLO('yolov8s.pt')  # load a pretrained model (recommended for training)
# Train the model
results = model1.train(data='data.yaml', epochs=50, imgsz=640)

metrics = model1.val()
metrics.box.map    # map50-95
metrics.box.map50  # map50
metrics.box.map75  # map75
metrics.box.maps   # a list contains map50-95 of each category

New https://pypi.org/project/ultralytics/8.1.11 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
engine\trainer: task=detect, mode=train, model=yolov8s.pt, data=data.yaml, epochs=50, patience=50, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train4, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, show=False, save_frames=False, save_txt=False, save_conf=False, save_c

train: Scanning C:\Users\rafae\Documents\Capstone\datasets\train\labels.cache... 63 images, 11 backgrounds, 0 corrupt: 100%|██████████| 63/63 [00:00<?, ?it/s]
val: Scanning C:\Users\rafae\Documents\Capstone\datasets\valid\labels.cache... 19 images, 1 backgrounds, 0 corrupt: 100%|██████████| 19/19 [00:00<?, ?it/s]


Plotting labels to runs\detect\train4\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs\detect\train4
Starting training for 50 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/50         0G      2.293      3.982      1.472        121        640: 100%|██████████| 4/4 [02:32<00:00, 38.10s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.88s/it]

                   all         19        151      0.424       0.26      0.278      0.134



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/50         0G      2.033       2.42      1.359        231        640: 100%|██████████| 4/4 [02:20<00:00, 35.13s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.97s/it]

                   all         19        151      0.433      0.488      0.398      0.203



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/50         0G      1.896      1.929      1.296        155        640: 100%|██████████| 4/4 [02:55<00:00, 43.82s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:18<00:00, 18.30s/it]

                   all         19        151      0.387      0.422      0.352      0.187



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/50         0G      1.737      1.594      1.249        118        640: 100%|██████████| 4/4 [02:50<00:00, 42.74s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:19<00:00, 19.70s/it]

                   all         19        151      0.315      0.469      0.323      0.162



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/50         0G      1.752      1.465      1.271        189        640: 100%|██████████| 4/4 [02:58<00:00, 44.57s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.50s/it]

                   all         19        151      0.347      0.417      0.348      0.193



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/50         0G      1.733      1.385      1.232        213        640: 100%|██████████| 4/4 [02:44<00:00, 41.10s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:17<00:00, 17.20s/it]

                   all         19        151      0.375      0.475      0.364       0.21



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/50         0G      1.634      1.341       1.21        133        640: 100%|██████████| 4/4 [02:56<00:00, 44.04s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:17<00:00, 17.14s/it]

                   all         19        151      0.369      0.516      0.345      0.183



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/50         0G      1.649      1.266      1.201        194        640: 100%|██████████| 4/4 [02:30<00:00, 37.57s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.61s/it]

                   all         19        151      0.405      0.608      0.361       0.21



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/50         0G      1.602      1.304      1.167        129        640: 100%|██████████| 4/4 [02:23<00:00, 35.84s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.34s/it]

                   all         19        151      0.402      0.504       0.38      0.217



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/50         0G      1.622      1.308      1.221         91        640: 100%|██████████| 4/4 [02:17<00:00, 34.35s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.23s/it]

                   all         19        151      0.571      0.421      0.438      0.252



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/50         0G      1.637      1.239      1.169        171        640: 100%|██████████| 4/4 [02:19<00:00, 34.84s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.41s/it]

                   all         19        151      0.533      0.486      0.461      0.268



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/50         0G      1.637       1.21      1.189        176        640: 100%|██████████| 4/4 [02:16<00:00, 34.18s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.62s/it]

                   all         19        151      0.555      0.471      0.493      0.295



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/50         0G      1.607      1.167      1.195        193        640: 100%|██████████| 4/4 [02:21<00:00, 35.41s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.98s/it]

                   all         19        151      0.504      0.587      0.544      0.299



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/50         0G       1.61      1.246      1.217        153        640: 100%|██████████| 4/4 [02:18<00:00, 34.68s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:17<00:00, 17.04s/it]

                   all         19        151      0.705       0.48      0.553      0.307



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/50         0G      1.515      1.134      1.172        154        640: 100%|██████████| 4/4 [02:21<00:00, 35.34s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:17<00:00, 17.72s/it]

                   all         19        151      0.535      0.563      0.525       0.27



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/50         0G      1.519      1.093      1.159        127        640: 100%|██████████| 4/4 [02:27<00:00, 36.92s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.53s/it]

                   all         19        151      0.608      0.482      0.533      0.288



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/50         0G      1.511       1.13      1.141        185        640: 100%|██████████| 4/4 [02:18<00:00, 34.54s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.59s/it]

                   all         19        151      0.617      0.463      0.503      0.289



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/50         0G      1.534      1.091      1.141        143        640: 100%|██████████| 4/4 [02:29<00:00, 37.47s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.36s/it]

                   all         19        151      0.409      0.591      0.476      0.278



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/50         0G      1.414      1.067      1.139        166        640: 100%|██████████| 4/4 [02:21<00:00, 35.41s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.53s/it]

                   all         19        151      0.444      0.513      0.446      0.262



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/50         0G      1.473      1.123      1.188        122        640: 100%|██████████| 4/4 [02:23<00:00, 35.84s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:19<00:00, 19.37s/it]

                   all         19        151      0.654      0.576      0.632      0.376



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/50         0G      1.488      1.026      1.179        153        640: 100%|██████████| 4/4 [03:05<00:00, 46.29s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.16s/it]

                   all         19        151      0.148      0.691      0.135     0.0708



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/50         0G      1.555      1.128      1.197        164        640: 100%|██████████| 4/4 [02:26<00:00, 36.53s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.98s/it]

                   all         19        151      0.741      0.607       0.68      0.365



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/50         0G      1.535      1.074      1.172        110        640: 100%|██████████| 4/4 [02:16<00:00, 34.22s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.66s/it]

                   all         19        151      0.717      0.619      0.651      0.383



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/50         0G      1.444      1.037      1.118        134        640: 100%|██████████| 4/4 [02:14<00:00, 33.60s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.70s/it]

                   all         19        151      0.753      0.524      0.654      0.392



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/50         0G      1.477      1.068       1.13        182        640: 100%|██████████| 4/4 [02:12<00:00, 33.12s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.69s/it]

                   all         19        151      0.689      0.584      0.664      0.404



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/50         0G      1.401     0.9948       1.12        138        640: 100%|██████████| 4/4 [02:30<00:00, 37.56s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:18<00:00, 18.96s/it]

                   all         19        151      0.683      0.588      0.698      0.439



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/50         0G      1.352      1.058      1.117        169        640: 100%|██████████| 4/4 [02:16<00:00, 34.18s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.23s/it]

                   all         19        151      0.803      0.534      0.711      0.446



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/50         0G      1.437      1.011      1.138        131        640: 100%|██████████| 4/4 [02:18<00:00, 34.74s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.34s/it]

                   all         19        151      0.838      0.586      0.731      0.466



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/50         0G      1.368     0.9987       1.11        156        640: 100%|██████████| 4/4 [02:16<00:00, 34.16s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.50s/it]

                   all         19        151      0.827      0.611      0.732      0.443



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/50         0G      1.338     0.9664      1.142        168        640: 100%|██████████| 4/4 [02:17<00:00, 34.46s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.82s/it]

                   all         19        151      0.806      0.616      0.751      0.465



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      31/50         0G      1.329     0.9262      1.074        181        640: 100%|██████████| 4/4 [02:07<00:00, 31.77s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.21s/it]

                   all         19        151       0.71      0.636      0.751      0.458



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      32/50         0G      1.351     0.9223      1.114        147        640: 100%|██████████| 4/4 [02:13<00:00, 33.41s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.53s/it]

                   all         19        151      0.736      0.619      0.739       0.45



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      33/50         0G      1.252     0.9052       1.07        163        640: 100%|██████████| 4/4 [02:15<00:00, 33.97s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.05s/it]

                   all         19        151      0.693      0.729      0.738      0.455



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      34/50         0G      1.283     0.8638       1.05        103        640: 100%|██████████| 4/4 [02:12<00:00, 33.03s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.26s/it]

                   all         19        151      0.652      0.718       0.72       0.46



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      35/50         0G        1.3     0.9319      1.133        187        640: 100%|██████████| 4/4 [02:13<00:00, 33.45s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.38s/it]

                   all         19        151      0.772      0.632      0.723      0.466



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      36/50         0G      1.232     0.8616      1.052        112        640: 100%|██████████| 4/4 [02:13<00:00, 33.47s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.67s/it]

                   all         19        151       0.74      0.648       0.73      0.478



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      37/50         0G      1.203     0.8357      1.067        115        640: 100%|██████████| 4/4 [02:11<00:00, 32.94s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.65s/it]

                   all         19        151      0.734      0.686       0.73      0.481



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      38/50         0G      1.179     0.8042      1.049        129        640: 100%|██████████| 4/4 [02:10<00:00, 32.58s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.36s/it]

                   all         19        151      0.711      0.781      0.766      0.501



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      39/50         0G      1.191     0.7819      1.076        131        640: 100%|██████████| 4/4 [02:15<00:00, 33.97s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.86s/it]

                   all         19        151      0.755      0.745      0.783      0.513



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      40/50         0G       1.18     0.7788      1.011        128        640: 100%|██████████| 4/4 [02:18<00:00, 34.64s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.94s/it]

                   all         19        151      0.758      0.666      0.766      0.477


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      41/50         0G      1.167     0.7878      1.033        120        640: 100%|██████████| 4/4 [02:11<00:00, 32.97s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.04s/it]

                   all         19        151      0.748      0.676      0.757      0.462



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      42/50         0G      1.155     0.7938      1.028         97        640: 100%|██████████| 4/4 [02:13<00:00, 33.27s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.24s/it]

                   all         19        151      0.741       0.72      0.777      0.482



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      43/50         0G      1.098      0.745      1.027         82        640: 100%|██████████| 4/4 [02:10<00:00, 32.74s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.83s/it]

                   all         19        151      0.775      0.786       0.79      0.502



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      44/50         0G       1.11      0.741      1.015         82        640: 100%|██████████| 4/4 [02:08<00:00, 32.07s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.84s/it]

                   all         19        151      0.743      0.798      0.793      0.508



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      45/50         0G      1.093     0.7213      1.003        109        640: 100%|██████████| 4/4 [02:15<00:00, 33.77s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.66s/it]

                   all         19        151      0.719      0.769      0.784      0.505



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      46/50         0G      1.084     0.7216      1.009        101        640: 100%|██████████| 4/4 [02:13<00:00, 33.26s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.34s/it]

                   all         19        151      0.793      0.672      0.788      0.497



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      47/50         0G      1.092     0.7285      1.025        104        640: 100%|██████████| 4/4 [02:10<00:00, 32.52s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:16<00:00, 16.02s/it]

                   all         19        151      0.798      0.706      0.792      0.507



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      48/50         0G      1.038     0.6891      0.987        130        640: 100%|██████████| 4/4 [02:09<00:00, 32.36s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.75s/it]

                   all         19        151      0.793      0.703        0.8      0.516



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      49/50         0G      1.052     0.6956     0.9969         99        640: 100%|██████████| 4/4 [02:10<00:00, 32.51s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.56s/it]

                   all         19        151      0.747      0.771      0.804      0.533



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      50/50         0G     0.9906     0.6784     0.9848         97        640: 100%|██████████| 4/4 [02:08<00:00, 32.16s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:15<00:00, 15.75s/it]

                   all         19        151      0.759      0.792      0.806      0.542



50 epochs completed in 2.220 hours.
Optimizer stripped from runs\detect\train4\weights\last.pt, 22.5MB
Optimizer stripped from runs\detect\train4\weights\best.pt, 22.5MB

Validating runs\detect\train4\weights\best.pt...
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 168 layers, 11126358 parameters, 0 gradients, 28.4 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:14<00:00, 14.60s/it]


                   all         19        151      0.759      0.792      0.806      0.542
                 Truck         19         26      0.612      0.808      0.775      0.488
                   car         19        125      0.906      0.776      0.838      0.596
Speed: 5.8ms preprocess, 742.4ms inference, 0.0ms loss, 0.8ms postprocess per image
Results saved to runs\detect\train4
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 168 layers, 11126358 parameters, 0 gradients, 28.4 GFLOPs


val: Scanning C:\Users\rafae\Documents\Capstone\datasets\valid\labels.cache... 19 images, 1 backgrounds, 0 corrupt: 100%|██████████| 19/19 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:14<00:00,  7.22s/it]


                   all         19        151      0.759      0.792      0.806      0.542
                 Truck         19         26      0.612      0.808      0.775      0.488
                   car         19        125      0.906      0.776      0.838      0.596
Speed: 5.8ms preprocess, 739.4ms inference, 0.0ms loss, 0.8ms postprocess per image
Results saved to runs\detect\train42


array([    0.48753,     0.59627])

![Description or alt text for the image](runs\detect\train4\results.png "Optional title")

### YOLOv8m (Medium)

In [6]:
model2 = YOLO('yolov8m.pt')  # load a pretrained model (recommended for training)
# Train the model
results = model2.train(data='data.yaml', epochs=50, imgsz=640)

metrics = model2.val()
metrics.box.map    # map50-95
metrics.box.map50  # map50
metrics.box.map75  # map75
metrics.box.maps   # a list contains map50-95 of each category

New https://pypi.org/project/ultralytics/8.1.11 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
engine\trainer: task=detect, mode=train, model=yolov8m.pt, data=data.yaml, epochs=50, patience=50, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train6, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, show=False, save_frames=False, save_txt=False, save_conf=False, save_c

train: Scanning C:\Users\rafae\Documents\Capstone\object_detect-1\train\labels... 63 images, 11 backgrounds, 0 corrupt: 100%|██████████| 63/63 [00:00<00:00, 122.25it/s]

train: New cache created: C:\Users\rafae\Documents\Capstone\object_detect-1\train\labels.cache



val: Scanning C:\Users\rafae\Documents\Capstone\object_detect-1\valid\labels... 19 images, 1 backgrounds, 0 corrupt: 100%|██████████| 19/19 [00:00<00:00, 107.12it/s]

val: New cache created: C:\Users\rafae\Documents\Capstone\object_detect-1\valid\labels.cache


Plotting labels to runs\detect\train6\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 77 weight(decay=0.0), 84 weight(decay=0.0005), 83 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs\detect\train6
Starting training for 50 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/50         0G      2.188       3.29      1.496        121        640: 100%|██████████| 4/4 [06:50<00:00, 102.64s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:36<00:00, 36.16s/it]

                   all         19        151      0.478      0.378      0.324      0.154



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/50         0G      1.963      1.955      1.372        231        640: 100%|██████████| 4/4 [06:41<00:00, 100.36s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/1 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 1.450s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:35<00:00, 35.45s/it]

                   all         19        151      0.394      0.107     0.0955     0.0526



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/50         0G      1.812      1.636      1.312        155        640: 100%|██████████| 4/4 [05:43<00:00, 85.75s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/1 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 1.450s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:35<00:00, 35.50s/it]

                   all         19        151      0.365      0.103     0.0961     0.0499



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/50         0G      1.705      1.622      1.289        118        640: 100%|██████████| 4/4 [05:11<00:00, 77.76s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.45s/it]

                   all         19        151      0.426      0.651      0.578       0.32



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/50         0G      1.706       1.45      1.253        189        640: 100%|██████████| 4/4 [05:44<00:00, 86.06s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.42s/it]

                   all         19        151      0.613      0.432      0.548      0.292



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/50         0G      1.676      1.348      1.231        213        640: 100%|██████████| 4/4 [05:48<00:00, 87.20s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:31<00:00, 31.39s/it]

                   all         19        151      0.543      0.643      0.646      0.358



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/50         0G      1.628      1.357      1.237        133        640: 100%|██████████| 4/4 [07:03<00:00, 105.75s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:43<00:00, 43.17s/it]

                   all         19        151      0.752      0.647      0.711      0.357



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/50         0G      1.602      1.345      1.241        194        640: 100%|██████████| 4/4 [06:25<00:00, 96.31s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:45<00:00, 45.85s/it]

                   all         19        151      0.694      0.474      0.554      0.309



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/50         0G      1.643      1.469      1.235        129        640: 100%|██████████| 4/4 [06:30<00:00, 97.59s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:36<00:00, 36.25s/it]

                   all         19        151      0.626      0.542      0.569      0.322



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/50         0G        1.6      1.486        1.3         91        640: 100%|██████████| 4/4 [06:06<00:00, 91.68s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:41<00:00, 41.64s/it]

                   all         19        151       0.66      0.537      0.526      0.268



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/50         0G        1.6      1.404       1.22        171        640: 100%|██████████| 4/4 [05:55<00:00, 88.81s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:32<00:00, 32.71s/it]

                   all         19        151      0.602      0.488       0.52      0.283



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/50         0G      1.601      1.307      1.212        176        640: 100%|██████████| 4/4 [05:14<00:00, 78.65s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:34<00:00, 34.05s/it]

                   all         19        151      0.608      0.351      0.389      0.213



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/50         0G      1.627      1.377      1.233        193        640: 100%|██████████| 4/4 [05:58<00:00, 89.55s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:34<00:00, 34.25s/it]

                   all         19        151      0.796      0.477       0.59      0.359



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/50         0G      1.615      1.314      1.221        153        640: 100%|██████████| 4/4 [05:54<00:00, 88.54s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:33<00:00, 33.59s/it]

                   all         19        151      0.578      0.484      0.519      0.304



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/50         0G      1.502      1.163      1.205        154        640: 100%|██████████| 4/4 [05:33<00:00, 83.43s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:34<00:00, 34.04s/it]

                   all         19        151      0.708      0.494      0.565      0.345



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/50         0G      1.517      1.067      1.194        127        640: 100%|██████████| 4/4 [05:17<00:00, 79.31s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:34<00:00, 34.20s/it]

                   all         19        151      0.633      0.505      0.478      0.275



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/50         0G      1.555      1.064      1.201        185        640: 100%|██████████| 4/4 [05:28<00:00, 82.17s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:32<00:00, 32.53s/it]

                   all         19        151      0.667      0.548      0.506      0.304



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/50         0G      1.527      1.035      1.179        143        640: 100%|██████████| 4/4 [06:05<00:00, 91.40s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:33<00:00, 33.81s/it]

                   all         19        151      0.606      0.531      0.543      0.318



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/50         0G      1.564      1.126      1.196        166        640: 100%|██████████| 4/4 [05:36<00:00, 84.00s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:30<00:00, 30.23s/it]

                   all         19        151      0.721      0.576      0.634      0.349



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/50         0G      1.495      1.099      1.208        122        640: 100%|██████████| 4/4 [05:56<00:00, 89.16s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:34<00:00, 34.41s/it]

                   all         19        151      0.716      0.577      0.625      0.333



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/50         0G       1.42          1      1.183        153        640: 100%|██████████| 4/4 [05:37<00:00, 84.29s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:31<00:00, 31.85s/it]

                   all         19        151      0.748      0.527      0.634      0.377



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/50         0G      1.447      1.091      1.178        164        640: 100%|██████████| 4/4 [05:03<00:00, 75.94s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:33<00:00, 33.83s/it]

                   all         19        151      0.667      0.643      0.632      0.383



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/50         0G      1.436       1.03      1.158        110        640: 100%|██████████| 4/4 [05:56<00:00, 89.21s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:33<00:00, 33.84s/it]

                   all         19        151      0.782      0.654      0.695       0.43



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/50         0G      1.515      1.016      1.171        134        640: 100%|██████████| 4/4 [06:03<00:00, 90.80s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:36<00:00, 36.84s/it]

                   all         19        151      0.195      0.525      0.183      0.104



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/50         0G      1.427      1.024      1.152        182        640: 100%|██████████| 4/4 [06:16<00:00, 94.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:35<00:00, 35.87s/it]

                   all         19        151      0.121      0.459      0.109     0.0631



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/50         0G      1.429     0.9864      1.144        138        640: 100%|██████████| 4/4 [06:01<00:00, 90.28s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:38<00:00, 38.19s/it]

                   all         19        151      0.196      0.496      0.176      0.108



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/50         0G      1.416     0.9295      1.163        169        640: 100%|██████████| 4/4 [06:29<00:00, 97.47s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.14s/it]

                   all         19        151      0.323      0.608      0.326        0.2



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/50         0G      1.363     0.9306      1.152        131        640: 100%|██████████| 4/4 [06:28<00:00, 97.14s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.51s/it]

                   all         19        151      0.782      0.643      0.706      0.445



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/50         0G      1.343     0.8939       1.14        156        640: 100%|██████████| 4/4 [06:03<00:00, 90.86s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.43s/it]

                   all         19        151      0.337      0.633      0.302      0.174



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/50         0G      1.278     0.9312      1.135        168        640: 100%|██████████| 4/4 [07:25<00:00, 111.27s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:35<00:00, 35.78s/it]

                   all         19        151      0.808      0.673      0.759      0.455



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      31/50         0G      1.333      0.854      1.108        181        640: 100%|██████████| 4/4 [06:09<00:00, 92.25s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:36<00:00, 36.83s/it]

                   all         19        151       0.76      0.756      0.805      0.493



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      32/50         0G      1.252     0.8535      1.097        147        640: 100%|██████████| 4/4 [06:25<00:00, 96.32s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.28s/it]

                   all         19        151      0.882      0.702       0.83      0.538



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      33/50         0G      1.224     0.8069      1.103        163        640: 100%|██████████| 4/4 [06:10<00:00, 92.57s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:39<00:00, 39.56s/it]

                   all         19        151      0.912      0.732      0.828      0.572



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      34/50         0G      1.211     0.7871      1.062        103        640: 100%|██████████| 4/4 [06:34<00:00, 98.58s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.99s/it]

                   all         19        151      0.902      0.725       0.83      0.569



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      35/50         0G      1.195     0.8316      1.121        187        640: 100%|██████████| 4/4 [07:15<00:00, 108.79s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:34<00:00, 34.82s/it]

                   all         19        151      0.876      0.742      0.855      0.591



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      36/50         0G      1.179     0.7776      1.059        112        640: 100%|██████████| 4/4 [06:21<00:00, 95.42s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:35<00:00, 35.59s/it]

                   all         19        151      0.816      0.755      0.848      0.606



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      37/50         0G      1.126     0.7512      1.077        115        640: 100%|██████████| 4/4 [06:07<00:00, 91.91s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:36<00:00, 36.81s/it]

                   all         19        151      0.883      0.792      0.862      0.617



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      38/50         0G      1.155     0.7383       1.08        129        640: 100%|██████████| 4/4 [06:11<00:00, 92.97s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.11s/it]

                   all         19        151      0.839      0.788      0.859      0.603



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      39/50         0G      1.134     0.7288      1.083        131        640: 100%|██████████| 4/4 [05:59<00:00, 89.77s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:36<00:00, 36.14s/it]

                   all         19        151      0.845      0.796      0.866      0.623



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      40/50         0G      1.104     0.7019      1.026        128        640: 100%|██████████| 4/4 [06:20<00:00, 95.07s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:35<00:00, 35.63s/it]

                   all         19        151        0.9      0.776      0.872       0.61


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      41/50         0G      1.052     0.7285       1.02        120        640: 100%|██████████| 4/4 [06:01<00:00, 90.36s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:35<00:00, 35.57s/it]

                   all         19        151      0.891      0.761      0.867      0.609



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      42/50         0G      1.073      0.696      1.027         97        640: 100%|██████████| 4/4 [06:22<00:00, 95.55s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.20s/it]

                   all         19        151       0.91      0.776      0.858      0.605



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      43/50         0G      0.997     0.6657      1.013         82        640: 100%|██████████| 4/4 [06:36<00:00, 99.00s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:39<00:00, 39.12s/it]

                   all         19        151      0.909      0.788      0.856      0.605



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      44/50         0G      1.037      0.664      1.016         82        640: 100%|██████████| 4/4 [06:13<00:00, 93.33s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:35<00:00, 35.19s/it]

                   all         19        151      0.903      0.784       0.85      0.598



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      45/50         0G     0.9808     0.6463      0.984        109        640: 100%|██████████| 4/4 [07:03<00:00, 105.92s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.10s/it]

                   all         19        151      0.919      0.765      0.864      0.602



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      46/50         0G     0.9707     0.6403     0.9759        101        640: 100%|██████████| 4/4 [06:24<00:00, 96.16s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:36<00:00, 36.81s/it]

                   all         19        151      0.838      0.811       0.87      0.614



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      47/50         0G     0.9909     0.6324      1.013        104        640: 100%|██████████| 4/4 [05:44<00:00, 86.12s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.38s/it]

                   all         19        151       0.83        0.8      0.876      0.622



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      48/50         0G     0.9085     0.6124     0.9613        130        640: 100%|██████████| 4/4 [06:14<00:00, 93.55s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.22s/it]

                   all         19        151      0.847       0.82      0.869      0.631



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      49/50         0G     0.9557     0.6409     0.9828         99        640: 100%|██████████| 4/4 [06:34<00:00, 98.62s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:37<00:00, 37.88s/it]

                   all         19        151      0.863       0.81      0.873      0.627



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      50/50         0G     0.8885     0.6243     0.9643         97        640: 100%|██████████| 4/4 [06:00<00:00, 90.07s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:34<00:00, 34.98s/it]

                   all         19        151      0.869       0.82      0.887      0.632



50 epochs completed in 5.668 hours.
Optimizer stripped from runs\detect\train6\weights\last.pt, 52.0MB
Optimizer stripped from runs\detect\train6\weights\best.pt, 52.0MB

Validating runs\detect\train6\weights\best.pt...
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 218 layers, 25840918 parameters, 0 gradients, 78.7 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:35<00:00, 35.92s/it]


                   all         19        151      0.869       0.82      0.887      0.632
                 Truck         19         26      0.844       0.83      0.902      0.656
                   car         19        125      0.894      0.811      0.873      0.607
Speed: 5.8ms preprocess, 1865.0ms inference, 0.0ms loss, 0.8ms postprocess per image
Results saved to runs\detect\train6
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 218 layers, 25840918 parameters, 0 gradients, 78.7 GFLOPs


val: Scanning C:\Users\rafae\Documents\Capstone\object_detect-1\valid\labels.cache... 19 images, 1 backgrounds, 0 corrupt: 100%|██████████| 19/19 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:35<00:00, 17.85s/it]


                   all         19        151      0.869       0.82      0.887      0.632
                 Truck         19         26      0.844       0.83      0.902      0.656
                   car         19        125      0.894      0.811      0.873      0.607
Speed: 6.6ms preprocess, 1855.0ms inference, 0.0ms loss, 0.8ms postprocess per image
Results saved to runs\detect\train62


array([    0.65557,     0.60745])

![Description or alt text for the image](runs\detect\train6\results.png "Optional title")


## 2.3 Models Comparsion

![Description or alt text for the image](yolov8.png "Optional title")

Training Time:

    -Nano: 1.145 hours
    -Small: 2.220 hours
    -Medium: 5.668 hours
    
Based on the training results and the given table for YOLOv8n (nano), YOLOv8s (small), and YOLOv8m (medium) models, here are some comparisons and considerations, especially in the context of a small dataset of only 100 images:

### Model Performance

**Accuracy:** There's a clear trend where the size of the model correlates with accuracy. The medium model has the highest mAP, followed by the small model, and then the nano model. However, the difference in mAP between the small and medium models is not as pronounced as between nano and small.

**Training Time:** As expected, the larger the model, the longer the training time. The nano model is the fastest to train, with the medium taking significantly longer.

### Loss and Metrics Trends

The training and validation loss for all models decrease over time, which is a good sign of learning. However, the larger models (small and medium) tend to have smoother and lower loss curves, indicating better performance.
The precision and recall metrics improve over time for all models. However, larger models tend to have higher precision and recall, indicating better overall predictive performance.

Larger models show higher mAP scores across both the 0.5 and 0.5-0.95 IoU thresholds, suggesting they are better at both detecting objects (mAP@0.5) and accurately predicting their shapes and sizes (mAP@0.5-0.95).

### Computational Resources and Time Constraints

**Nano Model:** Best if you have very strict time constraints and limited computational resources. It offers the least accuracy but is the fastest to train.

**Small Model:** Presents a balanced choice between accuracy and training time. It requires more time than the nano model but is significantly faster than the medium model, making it a suitable option if you have moderate resources.

**Medium Model:** While it offers the best accuracy, the training time is considerably longer. It may not be the best choice if you have limited computational resources or if the training time is a critical factor.

### Considering a Small Dataset

With a small dataset of 100 images, the risk of overfitting increases, especially with larger models. Regularization and data augmentation become more important to prevent this.
The benefit of larger models may not be as significant with a small dataset since they're designed to capture more complex patterns that may not be present in limited data.

## 2.4 Final Consideration

Given the limited dataset size and considering the balance between training time and accuracy, the small model (YOLOv8s) seems to be a good middle ground. It offers a substantial improvement in accuracy over the nano model for a reasonable increase in training time. Since the accuracy of the small and medium models is relatively close, the doubled training time for the medium model may not be justifiable, especially under limited computational resources and time constraints.

It's also worth noting that the increase in training time is not linear with the increase in model size. The medium model takes over twice as long to train as the small model, which doesn't correspond to a doubling in accuracy, indicating diminishing returns for the increased computational cost.

In summary, for your scenario, the YOLOv8s model would likely be the best option, offering a good trade-off between the increased accuracy it provides over the nano model and the significantly lower training time compared to the medium model.

## 3. Possible Model Fine-tuning Strategies

Based on the training results for the YOLOv8n (nano), YOLOv8s (small), and YOLOv8m (medium) models, we have the following recommendations for prototype fine-tuning:

### Dataset Expansion
- **Expand the dataset**: More images will help with generalization and reduce the risk of overfitting.

### Data Augmentation
- **Enhance data augmentation**: Introduce more variability to make the model robust, especially with smaller datasets.

### Regularization Techniques
- **Implement regularization**: Use techniques like dropout or weight decay to prevent overfitting.

### Hyperparameter Tuning
- **Optimize hyperparameters**: Systematically tune learning rate, batch size, and others, possibly using automated tools like Optuna or Hyperopt.

### Early Stopping and Patience
- **Early stopping**: Implement early stopping with a patience value (e.g., ~30 epochs) and adjust based on observed training behavior.

### Learning Rate Scheduling
- **Adjust learning rate schedule**: Experiment with reducing the learning rate on a plateau or cyclical learning rates.


In summary, the YOLOv8s model provides a balance between training efficiency and accuracy improvement over the nano model. As we scale our dataset and continue to fine-tune the model, maintaining a balance between model complexity and dataset size will be crucial.


## References

For further details and resources, please refer to the following:

- Ultralytics YOLO Models and Utilities: [Ultralytics GitHub](https://github.com/ultralytics/ultralytics)
- Canadian Adverse Driving Conditions Dataset: [CADC GitHub](https://github.com/asvath/cadcd)
- Video Object Detection Converter: [VOD-Converter GitHub](https://github.com/umautobots/vod-converter/blob/master/README.md)
- Getting Started with Training a YOLOv8 Object Detection Model: [Datature Blog](https://www.datature.io/blog/get-started-with-training-a-yolov8-object-detection-model)
